toolchain/check/decl_state.h

```diff
@@ -0,0 +1,80 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_CHECK_DECL_STATE_H_
+#define CARBON_TOOLCHAIN_CHECK_DECL_STATE_H_
+
+namespace Carbon::Check {
+
+class KeywordModifierSet {
+ public:
+  enum RawEnum {
+    // At most one of these access modifiers allowed for a given declaration,
+    // and if present it must be first:
+    Private = 1 << 0,
+    Protected = 1 << 1,
+
+    // At most one of these declaration modifiers allowed for a given
+    // declaration:
+    Abstract = 1 << 2,
+    Base = 1 << 3,
+    Default = 1 << 4,
+    Final = 1 << 5,
+    Impl = 1 << 6,
+    Virtual = 1 << 7,
+
+    // Sets of modifiers:
+    Access = Private | Protected,
+  };
+
+  KeywordModifierSet() : keywords(static_cast<RawEnum>(0)) {}
+  KeywordModifierSet(RawEnum k) : keywords(k) {}
+  auto Set(RawEnum to_set) const -> KeywordModifierSet {
+    KeywordModifierSet ret;
+    ret.keywords = static_cast<RawEnum>(keywords | to_set);
+    return ret;
+  }
+  auto SetPrivate() const -> KeywordModifierSet { return Set(Private); }
+  auto SetProtected() const -> KeywordModifierSet { return Set(Protected); }
+  auto SetAbstract() const -> KeywordModifierSet { return Set(Abstract); }
+  auto SetBase() const -> KeywordModifierSet { return Set(Base); }
+  auto SetDefault() const -> KeywordModifierSet { return Set(Default); }
+  auto SetFinal() const -> KeywordModifierSet { return Set(Final); }
+  auto SetImpl() const -> KeywordModifierSet { return Set(Impl); }
+  auto SetVirtual() const -> KeywordModifierSet { return Set(Virtual); }
+
+  auto Has(unsigned to_Has) const -> bool { return keywords & to_Has; }
+  auto HasPrivate() const -> bool { return Has(Private); }
+  auto HasProtected() const -> bool { return Has(Protected); }
+  auto HasAbstract() const -> bool { return Has(Abstract); }
+  auto HasBase() const -> bool { return Has(Base); }
+  auto HasDefault() const -> bool { return Has(Default); }
+  auto HasFinal() const -> bool { return Has(Final); }
+  auto HasImpl() const -> bool { return Has(Impl); }
+  auto HasVirtual() const -> bool { return Has(Virtual); }
+  auto GetRaw() const -> RawEnum { return keywords; }
+
+ private:
+  RawEnum keywords;
+};
+
+struct DeclState {
+  // FIXME: `Fn` or `Function`?
+  enum DeclKind { FileScope, Class, NamedConstraint, Fn, Interface, Let, Var };
+
+  DeclState(DeclKind k, Parse::Node f) : first_node(f), kind(k) {}
+  DeclState() : DeclState(FileScope, Parse::Node::Invalid) {}
```

Comment:
```suggestion
  explicit DeclState(DeclKind k, Parse::Node f) : first_node(f), kind(k) {}
  explicit DeclState() : DeclState(FileScope, Parse::Node::Invalid) {}
```

Style: we're generally marking all constructors explicit unless otherwise needed (unfortunately, doesn't look like this made it into https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/cpp_style_guide.md, I'll try to remember to do that as a followup)

Reply:
Done.